tests/docs/test_supported_confs.py

```diff
@@ -0,0 +1,493 @@
+        elif return_type == "DensityMatrix":
+
+            def cost_fn(x):
+                res = circuit(x)
+                probs = jnp.abs(res) ** 2
+                return probs[0][0]
+
+            # compute the gradient of the scalar cost function instead
+            # of the jacobian of the state directly - the latter is a
+            # separate test case
+            return jax.grad(cost_fn)(x)
+        else:
+            return jax.jacrev(circuit, holomorphic=complex)(x)
+    elif interface == "tf":
+        with tf.GradientTape() as tape:
+            out = circuit(x)
+        if return_type in grad_types:
+            return tape.gradient(out, [x])
+        elif return_type == State:
+
+            def cost_fn(x):
+                res = circuit(x)
+                probs = tf.math.abs(res) ** 2
+                return probs[0]
+
+            with tf.GradientTape() as tape:
+                out = cost_fn(x)
+
+            # compute the gradient of the scalar cost function instead
+            # of the jacobian of the state directly - the latter is a
+            # separate test case
+            return tape.gradient(out, [x])
+        elif return_type == "DensityMatrix":
+
+            def cost_fn(x):
+                res = circuit(x)
+                probs = tf.math.abs(res) ** 2
+                return probs[0][0]
+
+            with tf.GradientTape() as tape:
+                out = cost_fn(x)
+
+            # compute the gradient of the scalar cost function instead
+            # of the jacobian of the state directly - the latter is a
+            # separate test case
+            return tape.gradient(out, [x])
+        else:
+            return tape.jacobian(out, [x])
+    elif interface == "torch":
+        if return_type in grad_types:
+            res = circuit(x)
+            res.backward()
+            return x.grad
+        elif return_type == State:
+
+            def cost_fn(x):
+                res = circuit(x)
+                probs = torch.abs(res) ** 2
+                return probs[0]
+
+            # compute the gradient of the scalar cost function instead
+            # of the jacobian of the state directly - the latter is a
+            # separate test case
+            res = cost_fn(x)
+            res.backward()
+            return x.grad
+        elif return_type == "DensityMatrix":
+
+            def cost_fn(x):
+                res = circuit(x)
+                probs = torch.abs(res) ** 2
+                return probs[0][0]
+
+            # compute the gradient of the scalar cost function instead
+            # of the jacobian of the state directly - the latter is a
+            # separate test case
+            res = cost_fn(x)
+            res.backward()
+            return x.grad
+        else:
+            return F.jacobian(circuit, (x,))
+
+
+class TestSupportedConfs:
+    """Test that the supported configurations in the documentation
+    matches the supported configurations in the code"""
```

Comment:
Worth noting already here (and/or in the module docstring) that the main purpose of these tests is to uncover if some of the supported functionalities would raise errors. They do not necessarily test correctness.

Reply:
Added this clarification in the class docstring.